functions_advanced.ipynb

```diff
@@ -0,0 +1,4959 @@
+    "    - [Example 2: memoization](#Example-2:-memoization)\n",
+    "  - [Parametrized decorators](#Parametrized-decorators)\n",
+    "- [Generators](#Generators)\n",
+    "  - [Create an interable from a generator](#Create-an-interable-from-a-generator)\n",
+    "  - [Combining generators](#Combining-generators)\n",
+    "- [Exercises](#Exercises)\n",
+    "  - [Password checker factory](#Password-checker-factory)\n",
+    "  - [String range](#String-range)\n",
+    "  - [Read `n` lines](#Read-n-lines)\n",
+    "  - [Only run once](#Only-run-once)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "1e7a4298-0533-43e8-b7d3-2e330edac443",
+   "metadata": {},
+   "source": [
+    "We are going to cover the following topics:\n",
+    "\n",
+    "- Decorators\n",
+    "- Lambdas\n",
+    "- Arguments and object's mutability\n",
+    "- Generators"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2b7dac95-e83a-46c6-8ad5-6b55a369bf17",
+   "metadata": {},
+   "source": [
+    "## Recap"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "218acd08-b0f9-4c4f-83d3-613497d17b2e",
+   "metadata": {},
+   "source": [
+    "Before starting our deep dive on functions, we must revise quickly two important concepts. Have a look at the [Functions](./functions.ipynb#The-scope-of-a-function) notebook for more detail.\n",
+    "\n",
+    "1. Scopes and namespaces\n",
+    "2. Functions are objects"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b618d311-0e0d-4748-a434-39b5a0a0e585",
+   "metadata": {},
+   "source": [
+    "### Functions are objects"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "380773ce-09de-42c1-8146-2ed3502ddeca",
+   "metadata": {},
+   "source": [
+    "As an example, suppose that we want to create a \"password checker\", that is, a function that can verify if an input password complies with some rules (e.g., minumum length, a given number of special characters). We could create a function with the following signature:\n",
+    "\n",
+    "```python\n",
+    "def check_password(\n",
+    "    password: str,\n",
+    "    min_length: int,\n",
+    "    min_uppercase: int,\n",
+    "    min_punctuation:\n",
+    "    int, min_digits: int\n",
+    "    ) -> bool:\n",
+    "    \"\"\"Check if a given password complies with pre-defined rules.\"\"\"\n",
+    "```\n",
+    "\n",
+    "In various situations, passwords are subject to distinct rules. Once these rules are defined, our goal is to streamline the process of handling them. We aim to avoid repeatedly inputting them for every password to check, as this can become tedious."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "438649ea-35d6-4404-8250-fa76bc4fda9d",
+   "metadata": {},
+   "source": [
+    "We can instead define a so-called **higher-order function** (see [Functional programming](./functional_programming.ipynb#Higher-Order-Functions-/-Functions-as-Values): a function that returns another function.\n",
```

Comment:
I think I've introduced that bug, sorry :)
```suggestion
    "We can instead define a so-called **higher-order function** (see [Functional programming](./functional_programming.ipynb#Higher-Order-Functions-/-Functions-as-Values)): a function that returns another function.\n",
```

Reply:
Thanks 😉 BTW, links to anchors don't seem to work for me. The notebook opens at the beginning.

Reply:
For me, it worked in Jupyter Lab. Did you test it in VS code?